substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs

```diff
@@ -705,71 +710,109 @@ where
 	{
-		let mut stream_item = stream.next();
+		// make the stream abortable
+		let (stream, handle) = futures::stream::abortable(stream);
+		let cancel_handle = handle.clone();
 
-		// The stop event can be triggered by the chainHead logic when the pinned
-		// block guarantee cannot be hold. Or when the client is disconnected.
-		let connection_closed = sink.closed();
-		tokio::pin!(connection_closed);
-		let mut stop_event = futures_util::future::select(rx_stop, connection_closed);
+		// create a oneshot to propagate error messages
+		let (tx_send, mut tx_receive) = oneshot::channel();
 
-		while let Either::Left((Some(event), next_stop_event)) =
-			futures_util::future::select(stream_item, stop_event).await
-		{
-			let events = match event {
+		// TODO: is there a cleaner way to do this?
+		let tx_send = Arc::new(parking_lot::Mutex::new(tx_send));
+		let mut events = move |event| {
+			match event {
 				NotificationType::InitialEvents(events) => Ok(events),
 				NotificationType::NewBlock(notification) =>
 					self.handle_import_blocks(notification, &startup_point),
 				NotificationType::Finalized(notification) =>
 					self.handle_finalized_blocks(notification, &startup_point),
 				NotificationType::MethodResponse(notification) => Ok(vec![notification]),
-			};
-
-			let events = match events {
-				Ok(events) => events,
-				Err(err) => {
-					debug!(
-						target: LOG_TARGET,
-						"[follow][id={:?}] Failed to handle stream notification {:?}",
-						self.sub_id,
-						err
-					);
-					_ = sink.send(&FollowEvent::<String>::Stop).await;
-					return Err(err)
-				},
-			};
+			}
+			.map_err(|err| {
+				debug!(
+					target: LOG_TARGET,
+					"[follow][id={:?}] Failed to handle stream notification {:?}",
+					self.sub_id,
+					err
+				);
+				err
+			})
+		};
 
-			for event in events {
-				if let Err(err) = sink.send(&event).await {
-					// Failed to submit event.
-					debug!(
-						target: LOG_TARGET,
-						"[follow][id={:?}] Failed to send event {:?}", self.sub_id, err
-					);
-
-					let _ = sink.send(&FollowEvent::<String>::Stop).await;
-					// No need to propagate this error further, the client disconnected.
-					return Ok(())
+		let stream = stream
+			.map(move |event| events(event))
+			.then(move |event| {
+				let handle = handle.clone();
+				let tx_send = tx_send.clone();
+				async move {
+					let handle = handle.clone();
+					let tx_send = tx_send.clone();
+
+					let result = match event {
+						Ok(events) => stream::iter(events),
+						Err(err) => {
+							{
+								// TODO: this is very hacky
+								// grab the lock
+								// swap the channels
+								// return the error to the main channel
+								let mut tx_send = tx_send.lock();
+								let (mut s, _receiver) =
+									oneshot::channel::<SubscriptionManagementError>();
+								mem::swap(&mut s, &mut tx_send);
+								s.send(err).expect("shouldn't happpen.");
+							}
+							handle.abort();
```

Comment:
why not just do this only when the select! returns below?

Reply:
done